### PR TITLE
Number fixes

### PR DESCRIPTION
**Pull request by baylesj**
Found out today that our jsontestrunner hasn't been testing our CharReader, only the legacy Reader. I rushed out a fix for this, as part of cleaning up the logic for parsing numbers.

### CODE REVIEW

**Comment by dota17** [comment]
+1 from me.

File: src/lib_json/json_reader.cpp [hunk 1]
```diff
@@ -1587,8 +1582,8 @@ bool OurReader::decodeNumber(Token& token, Value& decoded) {
   if (isNegative) {
     // We use the same magnitude assumption here, just in case.
     const Value::UInt last_digit = value % 10;
-    decoded = -Value::LargestInt(value / 10) - last_digit;
-  } else if (value > Value::LargestUInt(Value::maxLargestInt)) {
+    decoded = -Value::LargestInt(value / 10) * 10 - last_digit;
+  } else if (value <= Value::LargestUInt(Value::maxLargestInt)) {
```
Comment (cdunn2001) [inline src/lib_json/json_reader.cpp]: Oh! This was a bug? Good catch.

Comment (dota17) [reply src/lib_json/json_reader.cpp]: Just wonder the bug, is there a example code to reproduce the bug?

File: src/jsontestrunner/main.cpp [hunk 2]
```diff
@@ -126,15 +127,39 @@ static int parseAndSaveValueTree(const Json::String& input,
...
-                                 Json::Value* root) {
-  Json::Reader reader(features);
-  bool parsingSuccessful =
-      reader.parse(input.data(), input.data() + input.size(), *root);
-  if (!parsingSuccessful) {
-    printf("Failed to parse %s file: \n%s\n", kind.c_str(),
-           reader.getFormattedErrorMessages().c_str());
-    return 1;
+                                 Json::Value* root, bool use_legacy) {
+  if (!use_legacy) {
```
Comment (dota17) [inline src/jsontestrunner/main.cpp]: Just like @kimsey0 mentioned in #1043 

If `CharReader` has some new features that `Reader` doesn't have.There will be some questions if we do not differentiate the test/data between `CharReader` and `Reader`

Comment (baylesj) [reply src/jsontestrunner/main.cpp]: Thanks for bringing this to my attention. My suggested fix in the new revision to to prepend "legacy_" to any tests that should also run on the legacy `Reader` class. New tests that break compat should not start with "legacy_".

**Comment by cdunn2001** [comment]
> Found out today that our jsontestrunner hasn't been testing our CharReader, only the legacy Reader.

I think we should prioritize that over all other work. And as implies @kimsey0 says in #1062, we really should separate the old Reader tests from CharReader tests, so we can alter them independently. We are more limited in the changes we can make to old Reader.

**Comment by cdunn2001** [comment]
I think this only needs clang-format.

**Comment by kimsey0** [comment]
@baylesj: Would you clang-format this?

**Comment by dota17** [comment]
@kimsey0 Done.

@baylesj if you don't need branch `number-fixes` anymore, please delete it.

**Pull request merged**